server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/StarTreeDocsFileManager.java

```diff
@@ -0,0 +1,296 @@
+     */
+    private void ensureDocumentReadable(int docId, boolean shouldCreateFileOutput) throws IOException {
+        try {
+            if (docId >= currentFileStartDocId && docId < numReadableStarTreeDocuments) {
+                return;
+            }
+            IOUtils.closeWhileHandlingException(starTreeDocsFileInput);
+            starTreeDocsFileInput = null;
+            if (docId < numStarTreeDocs) {
+                loadStarTreeDocumentFile(docId);
+            }
+            if (starTreeDocsFileInput != null) {
+                return;
+            }
+            closeAndMaybeCreateNewFile(shouldCreateFileOutput, numStarTreeDocs);
+            loadStarTreeDocumentFile(docId);
+        } catch (IOException ex) {
+            IOUtils.closeWhileHandlingException(this);
+            throw ex;
+        }
+    }
+
+    /**
+     * The fileToByteSizeMap is in the following format
+     * file1 == 521 [ contains docs from 0 to 520 ]
+     * file2 == 780 [ contains docs from 521 to 779 ]
+     * <p>
+     * This method loads the correct 'star-tree.documents' file based on the docId
+     * and updates the "currentFileStartDocId" and "numReadableStarTreeDocuments"
+     */
+    private void loadStarTreeDocumentFile(int docId) throws IOException {
+        int currentFileStartDocId = 0;
+        for (Map.Entry<String, Integer> entry : fileToEndDocIdMap.entrySet()) {
+            if (docId < entry.getValue()) {
+                starTreeDocsFileInput = tmpDirectory.openInput(entry.getKey(), state.context);
+                starTreeDocsFileRandomInput = starTreeDocsFileInput.randomAccessSlice(
+                    starTreeDocsFileInput.getFilePointer(),
+                    starTreeDocsFileInput.length() - starTreeDocsFileInput.getFilePointer()
+                );
+                numReadableStarTreeDocuments = entry.getValue();
+                break;
+            }
+            currentFileStartDocId = entry.getValue();
+        }
+        this.currentFileStartDocId = currentFileStartDocId;
+    }
+
+    /**
+     * This case handles when the requested document ID is beyond the range of the currently open 'star-tree.documents' file.
+     * In this scenario, the following steps are taken:
+     * <p>
+     * 1. Close the current 'star-tree.documents' file.
+     * 2. Create a new 'star-tree.documents' file if the operation involves appending new documents.
+     *    If the operation is only for reading existing documents, a new file is not created.
+     */
+    private void closeAndMaybeCreateNewFile(boolean shouldCreateFileForAppend, int numStarTreeDocs) throws IOException {
+        if (starTreeDocsFileOutput != null) {
+            IOUtils.close(starTreeDocsFileOutput);
+        }
+        currBytes = 0;
+        if (starTreeDocsFileOutput != null) {
+            fileToEndDocIdMap.put(starTreeDocsFileOutput.getName(), numStarTreeDocs);
+        }
+        if (shouldCreateFileForAppend) {
```

Comment:
Any time will this condition false ?

Reply:
This is for file format PR, when we need to read all the docs across all files to write it.